# <p style="text-align: justify;"><div class="alert alert-info" role="alert">Data curation</div></p>

## `Ana González Guerra` 

### `Data science master's student` 

## Index<a class="anchor" id="index"></a>
* [Loading data](#1)
* [Filtering missing values](#2)
    * [Physical predictors dataframe](#3)
    * [Genetic predictors dataframe](#4)
    * [Target variables dataframe](#5)
* [Dividing data into more manageable data frames](#6)
* [References](#ref)

## Loading data <a class="anchor" id ="1" ></a>

* [Returning to the index of contents](#index)

In [1]:
import pandas as pd
import numpy as np

We worked from two datasets which shared observations and were merged in order to have a higher number of observations.

### Loading old dataset

In [2]:
handle_old = pd.read_csv('DATOS_TFM_BRUTO.csv', delimiter = ",")

In [3]:
handle_old.head()

CODIGO   W1Fc  W1Ox  W1Lac  W1tFc  W1tOx  W1tLac  W1tBorg  W1t5Fc  \
0    GDCF007  109.0  92.0    1.9  133.0   95.0     2.3      8.0    59.0   
1    GDCF014   93.0  88.0    2.3  161.0   85.0     3.7     10.0   115.0   
2  KDCl01020   43.0  81.0    1.1   65.0   89.0     4.3     10.0    97.0   
3  KDCl01035  116.0  92.0    2.0   92.0   82.0     3.3      8.0   139.0   
4    GDCF004   84.0  90.0    3.4  151.0   82.0     4.3      8.0    74.0   

   W1t5Ox  ...  rs8111989  rs8192678  rs824205  rs833070  rs884736  rs921943  \
0    90.0  ...        C:T        A:G       G:G       A:G       C:T       A:G   
1    85.0  ...        T:T        A:G       A:G       A:G       C:C       G:G   
2    80.0  ...        T:T        A:G       G:G       G:G       C:C       G:G   
3    98.0  ...        C:T        G:G       A:G       G:G       C:T       A:G   
4    97.0  ...        C:T        A:G       A:G       A:G       C:T       A:G   

   rs9469199  rs970547  rs9902453  rs9939609  
0        A:G       A:A        A:G        A:A  
1        G:G       A:G        G:G        A:A  
2        A:G       A:A        A:G        A:T  
3        A:G       A:G        G:G        A:A  
4        G:G       A:G        A:G        T:T  

[5 rows x 231 columns]

### Loading new dataset

In [4]:
handle_new = pd.read_csv('nuevos_datos_27_03_20.csv',delimiter=';',engine='python')
#engine = 'python' gives the possibility to work with csv files with a strange codification that normally python
#does not how to read


In [5]:
handle_new.head()

código     CODIGO   W1Fc  W1Ox  W1Lac  W1tFc  W1tOx  W1tLac  W1tBorg  \
0      SGCF6    GDCF007  109.0  92.0    1.9  133.0   95.0     2.3      8.0   
1     SGCF13    GDCF014   93.0  88.0    2.3  161.0   85.0     3.7     10.0   
2  KDCI01014  KDCl01014   43.0  81.0    1.1   65.0   89.0     4.3     10.0   
3  KDCI01037  KDCl01037  116.0  92.0    2.0   92.0   82.0     3.3      8.0   
4      SGCF4    GDCF004   84.0  90.0    3.4  151.0   82.0     4.3      8.0   

   W1t5Fc  ...  rs8111989  rs8192678  rs824205  rs833070  rs884736  rs921943  \
0    59.0  ...        C:T        A:G       G:G       A:G       C:T       A:G   
1   115.0  ...        T:T        A:G       A:G       A:G       C:C       G:G   
2    97.0  ...        C:C        A:G       A:G       A:A       C:C       G:G   
3   139.0  ...        C:T        A:A       G:G       A:G       C:T       G:G   
4    74.0  ...        C:T        A:G       A:G       A:G       C:T       A:G   

   rs9469199  rs970547  rs9902453  rs9939609  
0        A:G       A:A        A:G        A:A  
1        G:G       A:G        G:G        A:A  
2        G:G       A:A        A:A        T:T  
3        G:G       A:G        A:G        T:T  
4        G:G       A:G        A:G        T:T  

[5 rows x 276 columns]

For joining both files we need a commom key, for the particular case it will be 'CODIGO', thus the column 'código' will be deleted with the aim of avoiding confusions. 

In [6]:
handle_new = handle_new.drop(columns = ['código'])


The new dataset contains a column called 'numero' that it was used for make a count of the individuals as predictors were being measured. It can be eliminated, it does not apport meaning to the analysis.

In [7]:
handle_new = handle_new.drop(columns = ['numero'])

Taking into account that 'PerCad' predictor (hip circumference) , it is a specific predictor for women individuals and the number of women in the sample it is 39, we decided to delete this column and the 'Sexo' predictor (sex of the individual) due to the impossibility of make sub-samples of the original data for man and women with such a low number of observations.

In [8]:
handle_new[handle_new['Sexo'] == 1].shape #  is the coding for a woman individual

(39, 274)

In [9]:
handle_new = handle_new.drop(columns = ['PerCad','Sexo'])

In [10]:
handle_old = handle_old.drop(columns = ['Sexo'])

Now, we check if the rows at the end of the datasets are completely occupied with null values:

In [11]:
handle_old.tail()

CODIGO   W1Fc  W1Ox  W1Lac  W1tFc  W1tOx  W1tLac  W1tBorg  W1t5Fc  \
128  GDCF042   89.0  82.0    1.7  134.0   70.0     3.8     10.0    68.0   
129  GDCF001  102.0  99.0    3.4  175.0   91.0     3.9      8.0   132.0   
130      NaN    NaN   NaN    NaN    NaN    NaN     NaN      NaN     NaN   
131      NaN    NaN   NaN    NaN    NaN    NaN     NaN      NaN     NaN   
132      NaN    NaN   NaN    NaN    NaN    NaN     NaN      NaN     NaN   

     W1t5Ox  ...  rs8111989  rs8192678  rs824205  rs833070  rs884736  \
128    97.0  ...        T:T        A:G       G:G       G:G       C:C   
129    97.0  ...        T:T        G:G       G:G       A:G       C:T   
130     NaN  ...        NaN        NaN       NaN       NaN       NaN   
131     NaN  ...        NaN        NaN       NaN       NaN       NaN   
132     NaN  ...        NaN        NaN       NaN       NaN       NaN   

     rs921943  rs9469199  rs970547  rs9902453  rs9939609  
128       G:G        A:G       G:G        A:A        A:T  
129       G:G        G:G       A:G        A:G        A:T  
130       NaN        NaN       NaN        NaN        NaN  
131       NaN        NaN       NaN        NaN        NaN  
132       NaN        NaN       NaN        NaN        NaN  

[5 rows x 230 columns]

In [12]:
handle_new.tail()

CODIGO   W1Fc  W1Ox  W1Lac  W1tFc  W1tOx  W1tLac  W1tBorg  W1t5Fc  \
127    GDCF042   89.0  82.0    1.7  134.0   70.0     3.8     10.0    68.0   
128    GDCF001  102.0  99.0    3.4  175.0   91.0     3.9      8.0   132.0   
129  KDCl01036    NaN   NaN    NaN    NaN    NaN     NaN      NaN     NaN   
130        NaN    NaN   NaN    NaN    NaN    NaN     NaN      NaN     NaN   
131        NaN    NaN   NaN    NaN    NaN    NaN     NaN      NaN     NaN   

     W1t5Ox  ...  rs8111989  rs8192678  rs824205  rs833070  rs884736  \
127    97.0  ...        T:T        A:G       G:G       G:G       C:C   
128    97.0  ...        T:T        G:G       G:G       A:G       C:T   
129     NaN  ...        T:T        G:G       G:G       G:G       C:C   
130     NaN  ...        NaN        NaN       NaN       NaN       NaN   
131     NaN  ...        NaN        NaN       NaN       NaN       NaN   

     rs921943  rs9469199  rs970547  rs9902453  rs9939609  
127       G:G        A:G       G:G        A:A        A:T  
128       G:G        G:G       A:G        A:G        A:T  
129       A:G        G:G       A:G        A:G        A:T  
130       NaN        NaN       NaN        NaN        NaN  
131       NaN        NaN       NaN        NaN        NaN  

[5 rows x 272 columns]

Besides of rows with zero values along their length we have a row without values for physical predictors and only values for genetic predictors (row 129). It seems reasonable to delete this row plus the last three rows of both datasets (filled with null values).

In [13]:
handle_old = handle_old.iloc[:130,:]

In [14]:
handle_new = handle_new.iloc[:129,:]

After discarding these rows, we merge the data sets into one by selecting the rows present in both data sets, the ones that only the handle_old data set apports, the ones that only the handle_new data set apports, and the rows that both include.

In [15]:
df = handle_old.merge(handle_new, how='outer', indicator='union')#merge

In [16]:
df[df['union'] == 'both'] #selecting all the rows

CODIGO   W1Fc  W1Ox  W1Lac  W1tFc  W1tOx  W1tLac  W1tBorg  W1t5Fc  \
0    GDCF007  109.0  92.0    1.9  133.0   95.0     2.3      8.0    59.0   
1    GDCF014   93.0  88.0    2.3  161.0   85.0     3.7     10.0   115.0   
4    GDCF004   84.0  90.0    3.4  151.0   82.0     4.3      8.0    74.0   
5    GDCF036  138.0  98.0    1.2  192.0   97.0     4.2      8.0   137.0   
7    GDCF050  136.0  98.0    1.3  164.0   97.0     2.8     10.0   121.0   
..       ...    ...   ...    ...    ...    ...     ...      ...     ...   
125  GDCF053  146.0  97.0    2.7  135.0   81.0     5.1     10.0   162.0   
126  GDCF024  116.0  99.0    1.7  160.0   92.0     4.6      9.0   107.0   
127  GDCF070  100.0  97.0    1.5  160.0   99.0     3.1     10.0   124.0   
128  GDCF042   89.0  82.0    1.7  134.0   70.0     3.8     10.0    68.0   
129  GDCF001  102.0  99.0    3.4  175.0   91.0     3.9      8.0   132.0   

     W1t5Ox  ...  PerCin        FM1        FM2      %FFM1      %FFM2  \
0      90.0  ...    65.0   7.883768   8.432778  83.045661  81.864993   
1      85.0  ...    67.5  10.455818  12.498375  80.084157  76.193572   
4      97.0  ...    67.5   6.914187  10.373099  87.869847  81.801580   
5      98.0  ...    78.5  19.727433  21.178970  69.743201  67.516917   
7      97.0  ...    65.5   4.917710  10.003592  90.451048  80.575550   
..      ...  ...     ...        ...        ...        ...        ...   
125    98.0  ...    89.0  11.009325  11.615565  88.287952  87.643016   
126    90.0  ...    91.0   8.531546  13.933039  91.249697  85.709704   
127    99.0  ...    85.5   6.392673   9.594553  92.601073  88.895193   
128    97.0  ...    88.5   7.949812  12.728019  91.424151  86.269666   
129    97.0  ...    98.0  22.898710  20.736852  76.393082  78.621802   

          FFM1       FFM2      FFMI1      FFMI2  union  
0    38.616232  38.067222  16.714090  16.476464   both  
1    42.044182  40.001625  17.276538  16.437223   both  
4    50.085813  46.626901  17.330731  16.133876   both  
5    45.472567  44.021030  16.111312  15.597020   both  
7    46.582290  41.496408  17.749691  15.811770   both  
..         ...        ...        ...        ...    ...  
125  82.990675  82.384435  23.988517  23.813283   both  
126  88.968454  83.566961  27.767065  26.081259   both  
127  80.007327  76.805447  23.126179  22.200673   both  
128  84.750188  79.971981  24.762655  23.366539   both  
129  74.101290  76.263148  23.127022  23.801738   both  

[102 rows x 274 columns]

If we select the rows that are in both datasets we have a total of 102 rows.

In [17]:
df_merged = df[df['union'] == 'both'] 

## Filtering missing values <a class="anchor" id="2"></a>

* [Returning to the index of contents](#index)

In order to work always with the same null value format we will convert all the existing ones to np.nan:

In [18]:
df_merged = df_merged.replace({'No Call':np.nan, 'Invalid':np.nan, '-:-':np.nan})

Now, the null values have been standardised, we analyse their presence in our data with the function missing_values_table(df) [[1]](#koehrsen2018).

In [19]:
def missing_values_table(df):
        # Total missing values
        mis_val = df.isnull().sum()
        
        # Percentage of missing values
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        
        # Make a table with the results
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        
        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        
        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        
        # Print some summary information
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        
        # Return the dataframe with missing information
        return mis_val_table_ren_columns

In [20]:
col_mis_merged = missing_values_table(df_merged)#columns of the dataframe with missing values
col_mis_merged

Your selected dataframe has 274 columns.
There are 209 columns that have missing values.


Missing Values  % of Total Values
rs3025058               102              100.0
rs1800012.1              70               68.6
rs3213221.1              70               68.6
rs384129                 45               44.1
rs12722                  38               37.3
...                     ...                ...
W2MaxWAbs                 1                1.0
W2maxRPM                  1                1.0
W2MedWAbs                 1                1.0
W2medRPM                  1                1.0
W1t5Fc                    1                1.0

[209 rows x 2 columns]

We have a high number of columns with missing values, but not all the columns have a high number of missing values. Let's look at the columns with more than 5 missing values:

In [21]:
col_mis_merged[col_mis_merged['Missing Values'] >=5]

Missing Values  % of Total Values
rs3025058               102              100.0
rs1800012.1              70               68.6
rs3213221.1              70               68.6
rs384129                 45               44.1
rs12722                  38               37.3
rs1799750                27               26.5
rs8065080                 6                5.9
rs7498665                 6                5.9
rs7292407                 6                5.9
rs7965584                 5                4.9
rs3219178                 5                4.9
rs1801133                 5                4.9
rs1801260                 5                4.9
rs11715829                5                4.9
rs824205                  5                4.9
rs833070                  5                4.9
rs4654748                 5                4.9
rs6735530                 5                4.9

There is a total of 209 columns with missing values, it can be appreciated that columns with the highest number of missing values are the those referring to the SNPs. We know that some SNPs were replicated due to problems detecting genotypes, these SNPs are: rs1800012, rs3213221 and rs6265. 

Besides of these SNPs replicated, the SNP 'rs3025058' has a 100% of missing values, that does not contain any value, this column is directly removed:

In [22]:
df_merged = df_merged.drop(columns = ['rs3025058'])

Since there are other SNPs with a greater number of missing values, we will determine if there are more replicated SNPs, in addition to the ones mentioned above. To detect these possible replicas we draw from the premise that the columns that are a duplicate will have a name ending in '.1'.

In [23]:
snp = pd.DataFrame()
for i in df_merged:
    if i.endswith('.1'):
        snp = snp.append(df_merged[i])
snp = snp.T
print('Replicated SNPs:',snp.columns)

Replicated SNPs: Index(['rs1800012.1', 'rs3213221.1', 'rs6265.1'], dtype='object')


It seems that the only SNPs replicated are the initial ones mentioned above. With the aim of losing the lesser amount of information for these SNPs we will make a combination of the original and replica column into one column, based on the next criteria (these criteria are using the example of a predictor with possible genotypes: 'AA','AG','GG'):

* If we have 'AA' in one column, and 'GG' in the other, the result in the final column is 'AG'.
* If we have 'AG' in one column, and 'GG' in the other, the result in the final column is 'AG'.
* If we have 'AA' or 'GG' in both columns, the result in the final column is 'AA' or 'GG', respectively.
* If we have a null value in one column and an actual value in the other, we take the value of the column with actual value.


Possible genotypes for 'rs1800012':

In [24]:
df_merged['rs1800012'].unique()

array(['G:T', 'G:G', 'T:T', nan], dtype=object)

In [25]:
#SNP  'rs1800012' and replica 

for i in np.arange(len(df_merged)):
    
    if df_merged['rs1800012'].values[i] is np.nan and df_merged['rs1800012.1'].values[i] is not np.nan:
        df_merged['rs1800012'].values[i] = df_merged['rs1800012.1'].values[i]
    
    elif df_merged['rs1800012'].values[i] is not np.nan and df_merged['rs1800012.1'].values[i] is np.nan:
        pass
    
    elif df_merged['rs1800012'].values[i] == df_merged['rs1800012.1'].values[i]:
        pass
    
    elif df_merged['rs1800012'].values[i] is not df_merged['rs1800012.1'].values[i]:
        if df_merged['rs1800012'].values[i] is not np.nan and df_merged['rs1800012.1'].values[i] is not np.nan:
            df_merged['rs1800012'].values[i] = 'G:T'
    else:
        pass
    

        
        

Possible genotypes for 'rs3213221':

In [26]:
df_merged['rs3213221'].unique()

array(['C:C', 'C:G', 'G:G', nan], dtype=object)

In [27]:
#SNP 'rs3213221' and replica
for i in np.arange(len(df_merged)):
    
    if df_merged['rs3213221'].values[i] is np.nan and df_merged['rs3213221.1'].values[i] is not np.nan:
        df_merged['rs3213221'].values[i] = df_merged['rs3213221.1'].values[i]
    
    elif df_merged['rs3213221'].values[i] is not np.nan and df_merged['rs3213221.1'].values[i] is np.nan:
        pass
    
    elif df_merged['rs3213221'].values[i] == df_merged['rs3213221.1'].values[i]:
        pass
    
    elif df_merged['rs3213221'].values[i] is not df_merged['rs3213221.1'].values[i]:
        if df_merged['rs3213221'].values[i] is not np.nan and df_merged['rs3213221.1'].values[i] is not np.nan:
            df_merged['rs3213221'].values[i] = 'C:G'
    else:
        pass
    

Possible genotypes for 'rs6265':

In [28]:
df_merged['rs6265'].unique()

array(['A:G', 'G:G', nan, 'A:A'], dtype=object)

In [29]:
#SNP 'rs6265' and replica
for i in np.arange(len(df_merged)):
    
    if df_merged['rs6265'].values[i] is np.nan and df_merged['rs6265.1'].values[i] is not np.nan:
        df_merged['rs6265'].values[i] = df_merged['rs6265.1'].values[i]
    
    elif df_merged['rs6265'].values[i] is not np.nan and df_merged['rs6265.1'].values[i] is np.nan:
        pass
    
    elif df_merged['rs6265'].values[i] == df_merged['rs6265.1'].values[i]:
        pass
    
    elif df_merged['rs6265'].values[i] is not df_merged['rs6265.1'].values[i]:
        if df_merged['rs6265'].values[i] is not np.nan and df_merged['rs6265.1'].values[i] is not np.nan:
            df_merged['rs6265'].values[i] = 'A:G'
    else:
        pass
    

Now, we delete the replica columns:

In [30]:
df_merged = df_merged.drop(columns = ['rs1800012.1','rs3213221.1','rs6265.1'])

After having solved the replicas problem, we are going to look for other SNPs with a high number of missing values, with the aim to decide if they are included or not in the analysis.

In [31]:
missing_values_table(df_merged)

Your selected dataframe has 270 columns.
There are 205 columns that have missing values.


Missing Values  % of Total Values
rs384129               45               44.1
rs12722                38               37.3
rs1799750              27               26.5
rs7498665               6                5.9
rs7292407               6                5.9
...                   ...                ...
W2MaxWAbs               1                1.0
W2maxRPM                1                1.0
W2MedWAbs               1                1.0
W2medRPM                1                1.0
W1t5Fc                  1                1.0

[205 rows x 2 columns]

We decided to delete the columns with a percentage equal to or greater than 26.5% in missing values to avoid genetic predictors with few data or observations.

In [32]:
missing_df = missing_values_table(df_merged);
missing_columns = list(missing_df[missing_df['% of Total Values'] >= 26.5].index)
print('Columns removed %d.' % len(missing_columns))

df_merged= df_merged.drop(columns = list(missing_columns))

Your selected dataframe has 270 columns.
There are 205 columns that have missing values.
Columns removed 3.


Once we have eliminated the columns with the greatest number of missing values, we eliminate the rows that have at least one null value among their values. 

In [33]:
df_merged = df_merged.dropna(how='any') 

After this filtering we no longer have missing values:

In [34]:
missing_values_table(df_merged)

Your selected dataframe has 267 columns.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []

We check for the data dimension:

In [35]:
df_merged.shape

(82, 267)

Data has acceptable dimensions. From now we will work with this data.

But first, it is necessary to delete the 'CODIGO' column with the aim of deleting the individual identifiers, they are not essential for the modelling and we avoid the risk of compromising their personal identity.

In [36]:
df_merged = df_merged.drop(columns = ['CODIGO'])

Once the datasets have been joined, we no longer need the 'union' column, so we eliminate it.

In [37]:
df_merged= df_merged.drop(columns = ['union'])

We export the final dataset:

In [38]:
df_merged.to_csv('df_merged.csv',sep = ',')

## Dividing data into more manageable data frames <a class="anchor" id="6"></a>

* [Returning to the index of contents](#index)

### Physical predictors dataframe <a class="anchor" id="3"></a>

* [Returning to the index of contents](#index)

To construct the data framework for physical predictors, we exclude all target variables (those starting with "W1M", "W1m", "W2M", "W2m", "W1D", "W2D", "W1-2") and genetic predictors (those starting with "rs").

In [39]:
handle_physical = pd.DataFrame()
for i in df_merged:
    if i.lower().startswith('w1m') or  i.lower().startswith('w2m') or i.lower().startswith('w1d')or i.lower().startswith('w2d')or i.lower().startswith('w1-2'):
        pass
    elif i.startswith('rs'): 
        pass
    else:
        handle_physical = handle_physical.append(df_merged[i])
handle_physical = handle_physical.T
handle_physical

W1Fc  W1Ox  W1Lac  W1tFc  W1tOx  W1tLac  W1tBorg  W1t5Fc  W1t5Ox  \
0    109.0  92.0    1.9  133.0   95.0     2.3      8.0    59.0    90.0   
4     84.0  90.0    3.4  151.0   82.0     4.3      8.0    74.0    97.0   
7    136.0  98.0    1.3  164.0   97.0     2.8     10.0   121.0    97.0   
8    120.0  97.0    1.8  150.0   76.0     4.2      8.0   141.0    98.0   
9     62.0  97.0    2.8  172.0   91.0     4.9      9.0   119.0    88.0   
..     ...   ...    ...    ...    ...     ...      ...     ...     ...   
123  103.0  98.0    2.8  186.0   93.0     5.0     10.0   140.0    98.0   
124  159.0  99.0    2.4  181.0   86.0     6.0     10.0   149.0    95.0   
126  116.0  99.0    1.7  160.0   92.0     4.6      9.0   107.0    90.0   
127  100.0  97.0    1.5  160.0   99.0     3.1     10.0   124.0    99.0   
129  102.0  99.0    3.4  175.0   91.0     3.9      8.0   132.0    97.0   

     W1t5Lac  ...  PerCue  PerCin        FM1        FM2      %FFM1      %FFM2  \
0        4.7  ...    31.0    65.0   7.883768   8.432778  83.045661  81.864993   
4        6.3  ...    36.0    67.5   6.914187  10.373099  87.869847  81.801580   
7        9.0  ...    31.0    65.5   4.917710  10.003592  90.451048  80.575550   
8        9.7  ...    35.5    68.5  11.222758  14.984471  81.602036  75.435294   
9       10.8  ...    33.5    69.8  15.642411  16.063989  72.557174  71.817564   
..       ...  ...     ...     ...        ...        ...        ...        ...   
123     11.8  ...    41.0    80.5   5.348518   8.201681  93.142926  89.485024   
124     15.2  ...    46.0   110.0  28.244785  30.647354  74.554248  72.389772   
126     14.8  ...    46.5    91.0   8.531546  13.933039  91.249697  85.709704   
127     13.5  ...    43.5    85.5   6.392673   9.594553  92.601073  88.895193   
129     13.5  ...    44.0    98.0  22.898710  20.736852  76.393082  78.621802   

          FFM1       FFM2      FFMI1      FFMI2  
0    38.616232  38.067222  16.714090  16.476464  
4    50.085813  46.626901  17.330731  16.133876  
7    46.582290  41.496408  17.749691  15.811770  
8    49.777242  46.015529  18.507303  17.108689  
9    41.357589  40.936011  16.566892  16.398018  
..         ...        ...        ...        ...  
123  72.651482  69.798319  23.454120  22.533032  
124  82.755215  80.352646  25.541733  24.800200  
126  88.968454  83.566961  27.767065  26.081259  
127  80.007327  76.805447  23.126179  22.200673  
129  74.101290  76.263148  23.127022  23.801738  

[82 rows x 44 columns]

We check handle_physical columns:

In [40]:
handle_physical.columns

Index(['W1Fc', 'W1Ox', 'W1Lac', 'W1tFc', 'W1tOx', 'W1tLac', 'W1tBorg',
       'W1t5Fc', 'W1t5Ox', 'W1t5Lac', 'W2tFc', 'W2tOx', 'W2tLac', 'W2tBorg',
       'W2t5Fc', 'W2t5Ox', 'W2t5Lac', 'W2t10Fc', 'W2t10Ox', 'W2t10Lac',
       'Sumat', 'Peso', 'Edad', 'Altura', 'tpcgrasa', 'PlieguePec',
       'PliegueAx', 'PliegueTri', 'PliegueSub', 'PliegueAbd', 'PliegueSup',
       'PliegueMus', 'tpcgrasa1', 'tpcgrasa2', 'PerCue', 'PerCin', 'FM1',
       'FM2', '%FFM1', '%FFM2', 'FFM1', 'FFM2', 'FFMI1', 'FFMI2'],
      dtype='object')

We have three kinds of fat percentage but it should be two of them, we look for the error:

In [41]:
handle_physical[['tpcgrasa','tpcgrasa1','tpcgrasa2']]

tpcgrasa  tpcgrasa1  tpcgrasa2
0    16.954339  16.954339  18.135007
4    12.130153  12.130153  18.198420
7     9.548952   9.548952  19.424450
8    18.397964  18.397964  24.564706
9    27.442826  27.442826  28.182436
..         ...        ...        ...
123   6.857074   6.857074  10.514976
124  25.445752  25.445752  27.610228
126   8.750303   8.750303  14.290296
127   7.398927   7.398927  11.104807
129  23.606918  23.606918  21.378198

[82 rows x 3 columns]

'tpcgrasa' is the same as 'tpcgrasa1', so we eliminate 'tpcgrasa':

In [42]:
handle_physical = handle_physical.drop(columns = ['tpcgrasa'])

These are the physical predictors that will be considered for the modelling:

In [43]:
handle_physical.columns

Index(['W1Fc', 'W1Ox', 'W1Lac', 'W1tFc', 'W1tOx', 'W1tLac', 'W1tBorg',
       'W1t5Fc', 'W1t5Ox', 'W1t5Lac', 'W2tFc', 'W2tOx', 'W2tLac', 'W2tBorg',
       'W2t5Fc', 'W2t5Ox', 'W2t5Lac', 'W2t10Fc', 'W2t10Ox', 'W2t10Lac',
       'Sumat', 'Peso', 'Edad', 'Altura', 'PlieguePec', 'PliegueAx',
       'PliegueTri', 'PliegueSub', 'PliegueAbd', 'PliegueSup', 'PliegueMus',
       'tpcgrasa1', 'tpcgrasa2', 'PerCue', 'PerCin', 'FM1', 'FM2', '%FFM1',
       '%FFM2', 'FFM1', 'FFM2', 'FFMI1', 'FFMI2'],
      dtype='object')

In [44]:
print('Number of physical predictors:',len(handle_physical.columns))

Number of physical predictors: 43


We export physical dataframe:

In [45]:
handle_physical.to_csv('handle_physical_fusion_31_03_20.csv', sep=',') 

### Genetic predictors dataframe <a class="anchor" id="4"></a>

* [Returning to the index of contents](#index)

In [46]:
handle_genetic = pd.DataFrame()
for i in df_merged:
    if i.startswith('rs'): 
        handle_genetic = handle_genetic.append(df_merged[i])
    else:
        pass
        
handle_genetic = handle_genetic.T


These will be the genetic predictors considered for the modelling:

In [47]:
print(list(handle_genetic.columns))

['rs10034661', 'rs10063949', 'rs1024610', 'rs1024611', 'rs1042713', 'rs1045485', 'rs10491003', 'rs1049305', 'rs1049434', 'rs10497520', 'rs1052700', 'rs1057972', 'rs10741657', 'rs10921078', 'rs10999516', 'rs11051548', 'rs1107946', 'rs11096663', 'rs1110976', 'rs11144134', 'rs11225395', 'rs11254363', 'rs1137100', 'rs1137101', 'rs1144393', 'rs11549465', 'rs11715829', 'rs1205', 'rs12077871', 'rs12150660', 'rs12255372', 'rs1260326', 'rs12613181', 'rs12934922', 'rs13146355', 'rs1330363', 'rs13946', 'rs1420100', 'rs1420106', 'rs143383', 'rs1451462', 'rs146797694', 'rs1481012', 'rs1516797', 'rs15705', 'rs1572312', 'rs1589241', 'rs1676486', 'rs16881983', 'rs16924573', 'rs17685', 'rs17782313', 'rs1799722', 'rs1799865', 'rs1799907', 'rs1799983', 'rs1800012', 'rs1800169', 'rs1800206', 'rs1800255', 'rs1800497', 'rs1800562', 'rs1800629', 'rs1800795', 'rs1800796', 'rs1800872', 'rs1801131', 'rs1801133', 'rs1801198', 'rs1801260', 'rs1801282', 'rs1805086', 'rs1815739', 'rs182549', 'rs1862513', 'rs1906058

In [48]:
print('Number of physical predictors:',len(handle_genetic.columns))

Number of physical predictors: 185


We export handle_genetic dataframe:

In [49]:
handle_genetic.to_csv('handle_genetic_fusion_31_03_20.csv', sep=',') 

### Target variables dataframe <a class="anchor" id="5"></a>

* [Returning to the index of contents](#index)

In [50]:
handle_target = pd.DataFrame()
for i in df_merged:
    if i.lower().startswith('w1m') or  i.lower().startswith('w2m') or i.lower().startswith('w1d')or i.lower().startswith('w2d')or i.lower().startswith('w1-2'):
        handle_target = handle_target.append(df_merged[i])
    else:
        pass
handle_target = handle_target.T

These will be the target variables considered for the modelling:

In [51]:
handle_target.columns

Index(['W1MaxWAbs', 'W1MaxWRel', 'W1maxRPM', 'W1MedWAbs', 'W1MedWRel',
       'W1medRPM', 'W2MaxWAbs', 'W2MaxWRel', 'W2maxRPM', 'W2MedWAbs',
       'W2MedWRel', 'W2medRPM', 'W1MaxWRelFFM1', 'W1MaxWRelFFM2',
       'W1MedWRelFFM1', 'W1MedWRelFFM2', 'W1DifWAbs', 'W1DifWRel',
       'W1DifWRelFFM1', 'W1DifWRelFFM2', 'W2MaxWRelFFM1', 'W2MaxWRelFFM2',
       'W2MedWRelFFM1', 'W2MedWRelFFM2', 'W2DifWAbs', 'W2DifWRel',
       'W2DifWRelFFM1', 'W2DifWRelFFM2', 'W1-2MaxDifWAbs', 'W1-2MaxDifWRel',
       'W1-2MaxDifWRelFFM1', 'W1-2MaxDifWRelFFM2', 'W1-2MedDifWAbs',
       'W1-2MedDifWRel', 'W1-2MedDifWRelFFM1', 'W1-2MedDifWRelFFM2'],
      dtype='object')

In [52]:
print('Number of target variables:',len(handle_target.columns))

Number of target variables: 36


Some additional target variables were added to this set with the intention of studying the evolution of power during the Wingate test in relation to fat folds.

* Maximum power relative to  'Sumat' in Wingate 1: 'W1MaxWRelPli'

In [53]:
handle_target['W1MaxWRelPli'] = handle_target['W1MaxWAbs'] / handle_physical['Sumat']

* Mean power relative to 'Sumat' in Wingate 1: 'W1MedWRelPli'

In [54]:
handle_target['W1MedWRelPli'] = handle_target['W1MedWAbs'] / handle_physical['Sumat']

*  Maximum power relative to  'Sumat' in Wingate 2: W2MaxWRelPli

In [55]:
handle_target['W2MaxWRelPli'] = handle_target['W2MaxWAbs'] / handle_physical['Sumat']

* Mean power relative to 'Sumat' in Wingate 2: W2MedWRelPli

In [56]:
handle_target['W2MedWRelPli'] = handle_target['W2MedWAbs'] / handle_physical['Sumat']

In [57]:
handle_target.columns

Index(['W1MaxWAbs', 'W1MaxWRel', 'W1maxRPM', 'W1MedWAbs', 'W1MedWRel',
       'W1medRPM', 'W2MaxWAbs', 'W2MaxWRel', 'W2maxRPM', 'W2MedWAbs',
       'W2MedWRel', 'W2medRPM', 'W1MaxWRelFFM1', 'W1MaxWRelFFM2',
       'W1MedWRelFFM1', 'W1MedWRelFFM2', 'W1DifWAbs', 'W1DifWRel',
       'W1DifWRelFFM1', 'W1DifWRelFFM2', 'W2MaxWRelFFM1', 'W2MaxWRelFFM2',
       'W2MedWRelFFM1', 'W2MedWRelFFM2', 'W2DifWAbs', 'W2DifWRel',
       'W2DifWRelFFM1', 'W2DifWRelFFM2', 'W1-2MaxDifWAbs', 'W1-2MaxDifWRel',
       'W1-2MaxDifWRelFFM1', 'W1-2MaxDifWRelFFM2', 'W1-2MedDifWAbs',
       'W1-2MedDifWRel', 'W1-2MedDifWRelFFM1', 'W1-2MedDifWRelFFM2',
       'W1MaxWRelPli', 'W1MedWRelPli', 'W2MaxWRelPli', 'W2MedWRelPli'],
      dtype='object')

* Maximum fatigue rate between Wingates relative to 'Sumat': 'W1-W2MaxDifWRelPli'

In [58]:
handle_target['W1-W2MaxDifWRelPli'] = handle_target['W1MaxWRelPli'] - handle_target['W2MaxWRelPli']

* Mean fatigue rate between Wingates relative to 'Sumat': 'W1-W2MedDifWRelPli'

In [59]:
handle_target['W1-W2MedDifWRelPli'] = handle_target['W1MedWRelPli'] - handle_target['W2MedWRelPli']

In [60]:
handle_target.columns

Index(['W1MaxWAbs', 'W1MaxWRel', 'W1maxRPM', 'W1MedWAbs', 'W1MedWRel',
       'W1medRPM', 'W2MaxWAbs', 'W2MaxWRel', 'W2maxRPM', 'W2MedWAbs',
       'W2MedWRel', 'W2medRPM', 'W1MaxWRelFFM1', 'W1MaxWRelFFM2',
       'W1MedWRelFFM1', 'W1MedWRelFFM2', 'W1DifWAbs', 'W1DifWRel',
       'W1DifWRelFFM1', 'W1DifWRelFFM2', 'W2MaxWRelFFM1', 'W2MaxWRelFFM2',
       'W2MedWRelFFM1', 'W2MedWRelFFM2', 'W2DifWAbs', 'W2DifWRel',
       'W2DifWRelFFM1', 'W2DifWRelFFM2', 'W1-2MaxDifWAbs', 'W1-2MaxDifWRel',
       'W1-2MaxDifWRelFFM1', 'W1-2MaxDifWRelFFM2', 'W1-2MedDifWAbs',
       'W1-2MedDifWRel', 'W1-2MedDifWRelFFM1', 'W1-2MedDifWRelFFM2',
       'W1MaxWRelPli', 'W1MedWRelPli', 'W2MaxWRelPli', 'W2MedWRelPli',
       'W1-W2MaxDifWRelPli', 'W1-W2MedDifWRelPli'],
      dtype='object')

We export target variables dataframe:

In [61]:
handle_target.to_csv('handle_target_fusion_31_03_20.csv',sep = ',')

## References <a class="anchor" id="ref"></a>

* [Returning to the index of contents](#index)

[1] Koehrsen, W. (2018) Missing values. Available at: <a href="https://github.com/WillKoehrsen/machine-learning-project-walkthrough/blob/master/Machine Learning Project Part 1.ipynb">https://github.com/WillKoehrsen/machine-learning-project-walkthrough/blob/master/Machine Learning Project Part 1.ipynb</a> (Accessed: March 31, 2020).
<a class="anchor" id="koehrsen2018"></a>